hyperapp.d.ts

```diff
@@ -0,0 +1,128 @@
+export as namespace Hyperapp
+
+/** @namespace [Virtual DOM] */
+
+/** The Virtual DOM representation of an Element
+ *
+ * @memberOf [Virtual DOM]
+*/
+export interface VirtualNode<Data> {
+  tag: string
+  data: Data
+  children: VirtualNodeChild<{} | null>[]
+}
+
+/** In the Virtual DOM a Child could be either a Virtual Node or a string
+ *
+ * @memberOf [Virtual DOM]
+*/
+export type VirtualNodeChild<Data> = VirtualNode<Data> | string
+
+export interface Component<Data> {
+  /** A Component is a function that return a custom Virtual Node
+   *
+   * @memberOf [Virtual DOM]
+  */
+  (data: Data, children: VirtualNodeChild<{} | null>[]): VirtualNode<Data>
```

Comment:
[React](https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components) and [Vue.js](https://vuejs.org/v2/guide/components.html#Props) refer to `data` as `props`, should we follow the same convention?

Reply:
Could be a good idea! We'd need to change it in app and h too.

Reply:
Just to be complete, Angular refers to it as `inputs`, although in Angular's case, it's impossible for an input not to also become state inside the component after it's received. 

I'm also for renaming, especially since in Vue.js `data` refers to state inside the component.

Reply:
Give the names you want to. I am really bad at naming stuff. I followed the `API.md`

Reply:
@Swizz Looks good, but what about the tests? We can't merge without tests.

Reply:
@andrewiggins  Are you always among us? Ready to help me about tests?

Reply:
I'm here :) Checkout what I did in picodom/picodom#19 if you want a simple example of some tests. In summary, add some typescript files that demonstrates how you would expect the types to be used, and add a call to `tsc` in the test npm script to make sure they compile correctly.